### PR TITLE
feat: refactor and comments

### PR DESCRIPTION
**Pull request by Gwash3189**
### Description

Hey @paulegradie, have a read through the comments I've left. I've refactored a few things to be more idiomatic javascript.
Doing so will make things more maintainable in the long run, purely because we're leaning into the ecosystem. 
I'd be like writing C# like javascript. Doable, but difficult to do the longer you do it, as the native & regularly used tools would start getting in the way.

I've also refactored a few touch points with Prisma, as it has to be held in a specific way to not open too many connections with the database

tl;dr - classes are great! but we should avoid instance variables as they introduce hard-to-navigate coupling. DI would fix that in .NET, but DI isn't as smooth in JS/Node/Typescript. Leaning into the module system is a much smoother path.

### CODE REVIEW

File: src/lib/client/prisma.ts [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+    log:
+        env.NODE_ENV === 'development' ? ['query', 'error', 'warn'] : ['error']
+  })
+}
+
+if (client === undefined) {
+  client = createClient()
+}
+
+export const prisma = client
```
Comment (paulegradie) [inline src/lib/client/prisma.ts:19]: Wow - this is super weird. Let me see if I understand - 

1. `let` a client for later
2. Define a function to assign the client
3. On intial module reference / load create the client bc client === undefined
4. ? The next time something tries to import the client, it'll run the entire file and client will have been assigned?

I sense the design is exploiting the module system - yes/no?

File: src/lib/services/stripe/stripeWebhookEntryService.ts [hunk 2]
```diff
@@ -1,73 +1,72 @@
...
+    }
 
-            case StripeEventsCustomBullshit.PaymentIntentPaymentFailed:
-                const paymentIntentFailed = event.data.object;
-                // TODO: do some shit like send an email or feedback to the portal
-                break;
+    // Handle the event
+    switch (event.type) {
+        case StripeEventsCustomBullshit.PaymentIntentSucceeded:
+            const paymentIntentSucceeded = event.data.object as SuccessfulPayment; // TODO: as TYPE THIS SHIT PLZ;
```
Comment (paulegradie) [inline src/lib/services/stripe/stripeWebhookEntryService.ts:56]: Yeah righto I suppose we could just create our own types that point to the info we are looking for 👍

File: src/lib/controllers/stripe/StripeWebhookController.tsx [hunk 3]
```diff
@@ -2,29 +2,35 @@ import stripeBackendClient from "@/lib/client/stripe";
...
 // This is an unauthenticated stripe webhook controller
 // we'll add middleware on to this to reject any request that does not
 // contain the required headers from stripe
 class StripeWebhookController extends Controller {
-    private readonly stripeWebhookHandler: StripeWebhookHandler;
 
     constructor() {
         super();
-        this.stripeWebhookHandler = new StripeWebhookHandler(); // Can we use injectify-js or something to that event to inject dependencies? That would be nice (not sure about compatibility with the controllers
+        // we should avoid assigning instance variables like this.
```
Comment (paulegradie) [inline src/lib/controllers/stripe/StripeWebhookController.tsx:18]: Everything (as much as feasible) as functions? Thats a lot like how much python code is written. 

I'm not gonna lie - it does feel gross haha I actually never even liked it in python! But yeah its idiomatic and easier to test so I'm on board. 👍 I'll try and steer that way.

File: src/lib/controllers/base/AuthenticatedBaseController.ts [hunk 4]
```diff
@@ -1,18 +1,17 @@
...
 
 export class AuthenticatedBaseController extends Controller {
-    public readonly db: PrismaClient;
+    public readonly db: typeof prisma;
 
     constructor() {
         super();
 
-        this.db = prismaClient;
+        this.db = prisma;
```
Comment (paulegradie) [inline src/lib/controllers/base/AuthenticatedBaseController.ts:14]: I see that we'll remove this

File: src/lib/data/database/entity-stores/PrismaRepository.ts [hunk 5]
```diff
@@ -1,13 +1,24 @@
-import prismaClient from "@/lib/client/prisma";
+import { prisma } from "@/lib/client/prisma";
 import { DomainEntity, Model } from "@/types/sharedTypes";
 
+// getting a properly typed generic repo with Prisma is hard (i've tried)
```
Comment (paulegradie) [inline src/lib/data/database/entity-stores/PrismaRepository.ts:4]: You had me at `is hard (i've tried)`

This ended up being a learning exercise. But its provided an interesting contrast between js and c# that makes what you're saying a little easier to understand. 

I wasn't sure when I started if this was a thing in js, so I'm happy to ditch it. Thanks for your insights 🙏

**Review by paulegradie — APPROVED** [review]
Thanks for the comments on this - really really valuable stuff. Taking it all on board.

I'll merge this and apply the rest to the branch. Fortunately, I had the foresight to only define and not try and actually use these repositories.  😆 So I'll take your refactor, and delete the repositories.

**Comment by paulegradie** [comment]
I've spent a good chunk of time studying this PR so I'm happy to take this and apply to the base branch. Helps me learn 🙏

File: src/lib/client/prisma.ts (hunk 1 above)
Comment (Gwash3189) [reply src/lib/client/prisma.ts:19]: That's right. On first `import`/`require`, the file gets interpreted and run through Node, resulting in `client` and `prisma` being assigned. 

Each subsequent `import`/`require` doesn't re-evaluate the file but instead uses the values created on the initial `import`/`require` values.

It doesn't exploit the module system per se, as it's a valid and common use, but it's a way to implement a singleton in a module system like Node's that doesn't rely on OO or classes.

File: src/lib/services/stripe/stripeWebhookEntryService.ts (hunk 2 above)
Comment (Gwash3189) [reply src/lib/services/stripe/stripeWebhookEntryService.ts:56]: I had a dig and it `event.data.object` is just typed as `Object` which is the same as `Record<string, any>`. Not very useful from Stripes end, but typing it helps us a little. If we wanted to be _extra_ sure that the values existed at run time, we could run it through `zod`.
